[netdata] add feature for BR to request router role upgrade

abtink opened the pull request:
This commit adds a new mechanism to allow border routers to request
router role upgrade.

A border router which provides IP connectivity (either adding an
external route prefix or an on-mesh prefix with default route flag)
and is acting as a REED is eligible to request a router role upgrade
by sending an "Address Solicit" request to leader with status reason
`BorderRouterRequest`. This reason is used when the number of active
routers in the Thread mesh is above the threshold, and only if the
number of existing eligible BRs (determined from the network data)
that are acting as router is less than two. This mechanism  allows up
to two eligible border routers to request router role upgrade when
the number of routers is already above the threshold.

This commit also adds a new test-case `test_br_upgrade_router_role`
which validates the behavior of the new mechanism.

---

Related to [SPEC-1033](https://threadgroup.atlassian.net/jira/core/projects/SPEC/issues/SPEC-1033).

EskoDijk commented on src/core/thread/network_data_notifier.cpp:
```diff
@@ -152,6 +166,38 @@ void Notifier::HandleCoapResponse(Error aResult)
+        LogInfo("Requesting router role as BR when above upgrade threshold");
+    }
+
+    IgnoreError(Get<Mle::MleRouter>().BecomeRouter(ThreadStatusTlv::kBorderRouterRequst));
```
This line is outside the brackets of above 'if' statement - I think it should be inside. I.e. only request upgrade with this reason if  ActiveRouterCount >= UpgradeThreshold.

jwhui replied:
That makes sense. If the number of routers is less than router upgrade threshold, then this BR will attempt to upgrade to router using existing Thread 1.x mechanisms.

I think the more important consideration here is whether we want to apply some random delay. If there are a large number of BRs in the network, I could see a large number of requests being sent to the Leader at the same time whenever the network data is updated and the conditions to upgrade are satisfied.

abtink replied:
Thanks guys.

My idea is that as soon as we configured as an eligible BR:
- If number of active routers is below threshold, we don't wait for normal mechanism to upgrade to a router role and request it immediately (this gives more chance for BRs to start as router from beginning).
  - We can still request it with `kBorderRouterRequst` reason and leader is expected to treat as a normal request when we are below the active routers threshold.
- If number of active is above threshold, then we would count the number of existing BRs (from network data) and decide based on that whether to attempt to become router or not.

I understand it going somewhat beyond what we described in the JIRA (though it still using the new mechanism) but I feel this may be useful behavior.

Thoughts?

> I think the more important consideration here is whether we want to apply some random delay. If there are a large number of BRs in the network, I could see a large number of requests being sent to the Leader at the same time whenever the network data is updated and the conditions to upgrade are satisfied.

On adding some delay to send the the requests, I think that make sense. We may be able to re-use the timer in this module (probably need to maintain some sort of state). Will try to add this later in next push.

abtink replied:
Thinking more about this, decided to keep it simple so we use the new `kBorderRouterRequest` only when we are above the active router threshold (i.e., when we are below the threshold, the normal mechanism for router upgrade is used).

Added this in the new push. Also there is a randomly chosen delay (up to 10 second) before sending "Address Solicit" with this reason once device becomes eligible for such a request.

simonlingoogle replied:
> Thinking more about this, decided to keep it simple so we use the new kBorderRouterRequest only when we are above the active router threshold (i.e., when we are below the threshold, the normal mechanism for router upgrade is used).

Agree. Use normal mechanism whenever possible is more conservative, thus less risky.

EskoDijk commented on src/core/thread/network_data_notifier.cpp:
```diff
@@ -152,6 +166,38 @@ void Notifier::HandleCoapResponse(Error aResult)
+
+    VerifyOrExit(Get<Local>().CountBorderRouters(kAnyRole) > 0);
+
+    if (Get<RouterTable>().GetActiveRouterCount() >= Get<Mle::MleRouter>().GetRouterUpgradeThreshold())
```
Here it should probably also check that ActiveRouterCount < MAX_ROUTERS ?

jwhui replied:
I believe that check would be implicitly covered by the `VerifyOrExit()` line immediately below.

`Mle::kRouterUpgradeBorderRouterRequestThreshold` is currently set to `2`.

We could add a `static_assert` to ensure that `Mle::kRouterUpgradeBorderRouterRequestThreshold` is less than `Mle::kMaxRouters`.

EskoDijk replied:
But the CountBorderRouters(kRouterRoleOnly) function immediately below doesn't check anything about the number of active routers? Only counts devices providing network data that 'looks like a BR'.  So in a situation that there are 32 allocated Routers already, the BR device in REED role would request ADDR_SOL.req every time some network data change occurs. (If I read the code correctly.)

Adding the static_assert is a good idea.

jwhui replied:
Ah, yes. There should be a check against `kMaxRouters` somewhere.

abtink replied:
Yes. Good point. We need to exit if we are already at max routers (I guess it may be harmless to request but wasteful for sure). Will update it in the next push

EskoDijk reviewed (commented):
Left 2 review comments/questions

simonlingoogle commented on src/core/thread/network_data_notifier.cpp:
```diff
@@ -152,6 +169,75 @@ void Notifier::HandleCoapResponse(Error aResult)
+    // request router role upgrade with `kBorderRouterRequst`
+    // status.
+
+    VerifyOrExit(Get<Local>().CountBorderRouters(kAnyRole) > 0);
```
why don't we count the number of BR entries advertised by this device in Leader's Network Data?

abtink replied:
This does the same (`NetworkData::Local` conatains all the stuff this device wants to register in network data)
Also we want to try to become router before we may register the entries (to avoid re-registering again once becoming leader).

simonlingoogle replied:
I doubt if we can become Router before we may register the entries because the entries will be synchronized to Leader  very soon, meanwhile the new BR Router request will be sent after a few seconds. 

So, I think there is no notable difference if the local data can be synced to Leader.
Is it better to treat Leader Network Data as the single source of truth, which should also simplify calls to ScheduleRouterRoleUpgradeIfEligible.

abtink replied:
I agree. I think @EskoDijk has a very good point on this (related to leader behavior in accepting/rejecting such request). Changed this in the new push (we check `Leader` network data to see if the device's rloc16 is present in it and also it meets the requirement for being counted as br providing IP connectivity).

simonlingoogle replied:
Thanks @EskoDijk for the lucid elaboration. 
I think it's a general convention to use the Leader's Network Data as the SSOT, and use the Local Network Data as the intent, to avoid tricky issues.

EskoDijk replied:
Just a thought: to be precise, the BR would have to check that the Leader's network data indicates that the BR is "eligible for upgrade" (i.e. providing connectivity) and also its own local version of network data says that the BR is "eligible for upgrade".  Only then does it make sense to request for upgrade.    The Leader's version of network data may be outdated, or not matching the intent  ;-)

abtink replied:
@EskoDijk interesting and good point. It is a corner case, but I think it is easy to handle and support. I updated `IsEligibleForRouterRoleUpgradeAsBorderRouter()` to now check in both local network data and in leader network data.

simonlingoogle replied:
I think `IsEligibleForRouterRoleUpgradeAsBorderRouter` has handled these cases correctly.

simonlingoogle commented on tests/scripts/thread-cert/test_br_upgrade_router_role.py:
```diff
@@ -0,0 +1,156 @@
+BR3 = 5
+
+
+class SrpAutoStartMode(thread_cert.TestCase):
```
```suggestion
class BrUpgradeRouterRole(thread_cert.TestCase):
```

simonlingoogle commented on tests/scripts/thread-cert/test_br_upgrade_router_role.py:
```diff
@@ -0,0 +1,156 @@
+        br2.add_prefix('2001:dead:beef:2222::/64', 'paros')
+        br2.register_netdata()
+        self.simulator.go(15)
+        self.assertEqual(br2.get_state(), 'router')
```
how do we make sure the BR does not downgrade if the number of BRs exceed the `RouterDowngradeThreshold` ?

abtink replied:
If it downgrades (for some reason), we notice it from the `HandleNotifierEvent()` role change (role become `Child`) and can then send the request again if still eligible.

EskoDijk replied:
@abtink Maybe I misunderstand your above comment. But if it downgrades (due to the usual downgrade rules for Routers), and then upgrades again (due to the new BorderRouterRequest rule), then the first thing it would do after upgrading is downgrading again, right? So an infinite sequence of upgrade/downgrade actions would result which isn't helpful.

I would say that if a BR is eligible to send BorderRouterRequest, but it is already a Router, then it should avoid downgrading.   So that would need to be checked in the regular downgrade code.

abtink replied:
I see. No I missed that.  I agree we should ensure to add a check in downgrade code path to not allow a BR that is eligible for such as request to be downgraded (this would still be limited to two such BRs in network data).

abtink replied:
Added such a check in the new push (it is for now a separate commit to make it easier to read/review).

simonlingoogle commented on src/core/thread/network_data_notifier.cpp:
```diff
@@ -152,6 +169,75 @@ void Notifier::HandleCoapResponse(Error aResult)
+
+void Notifier::HandleTimeTick(void)
+{
+    if (mRouterRoleUpgradeTimeout > 0)
+    {
+        mRouterRoleUpgradeTimeout--;
+    }
+
+    if (mRouterRoleUpgradeTimeout == 0)
```
```suggestion
    if (mRouterRoleUpgradeTimeout > 0 && --mRouterRoleUpgradeTimeout == 0)
```
nit: a commonly used patter for HandleTimeTick.

abtink replied:
I prefer the existing one, since if we enter this method with `mRouterRoleUpgradeTimeout` being 0 already, we still want to go through the second `if`.

This should not happen with current use, but I feel it is better to behave correctly in this situation (in case code is changed in future which allows this pattern).

simonlingoogle replied:
> if we enter this method with `mRouterRoleUpgradeTimeout` being 0 already, we still want to go through the second `if`.

I think this is the very reason why `(mRouterRoleUpgradeTimeout > 0 && --mRouterRoleUpgradeTimeout == 0)` is a better pattern. 

`mRouterRoleUpgradeTimeout = 0` should actually mean that the timer is not set. So, we should not execute the timeout logic in such case. 

Thoughts?

abtink replied:
I think either way is fine. I changed it in the new push.

simonlingoogle commented on src/core/thread/network_data_notifier.cpp:
```diff
@@ -152,6 +169,75 @@ void Notifier::HandleCoapResponse(Error aResult)
+    return isEligible;
+}
+
+void Notifier::RequestRouterRoleIfEligible(void)
```
```suggestion
void Notifier::ScheduleRequestRouterRoleIfEligible(void)
```
since we only schedule it

simonlingoogle commented on src/core/thread/network_data_notifier.cpp:
```diff
@@ -107,6 +117,13 @@ void Notifier::HandleNotifierEvents(Events aEvents)
+#if OPENTHREAD_FTD && OPENTHREAD_CONFIG_BORDER_ROUTER_ENABLE && OPENTHREAD_CONFIG_BORDER_ROUTER_REQUEST_ROUTER_ROLE
+    if (aEvents.ContainsAny(kEventThreadRoleChanged | kEventThreadNetdataChanged))
+    {
+        RequestRouterRoleIfEligible();
```
If the leader does not support `kBorderRouterRequst` requests, we would keep sending requests in vain. Should we avoid that?

EskoDijk replied:
In case a legacy Leader does not CoAP-ACK the request (because it does not recognize the kBorderRouterRequest status), there are also CoAP retransmissions sent.  A legacy Leader could also ACK the request but not send the response. Or it could ACK the request and send a response status of 'No Address Available'.  All of these cases may have a different timeout/retransmission behavior. Good to keep in mind.

In any case if the ADDR_SOL.req message has been ACK'ed by the Leader the BR but no Success response then the BR can at least assume the Leader's not willing to grant the request and could stop sending these -- until the Leader/Partition changes.

If the ADDR_SOL.req message is not ACK'ed by the Leader then the BR may also stop sending (like above) just for consistency and treat it as a denial of the request.   (Even though officially in CoAP this situation means that the CoAP request was never received by the Leader.)

Another solution could be to only retry such request again when the number of Routers in the Partition has changed. I.e. use that as a simple trigger to avoid sending requests in vain whenever some minor thing in the network data is changing.   New Leader/Partition is in any case a trigger to send the request again.

abtink replied:
Thanks guys. I agree we need to avoid retrying.

In the new push, I added a mechanism to limit sending such a request to once while local network data remains unchanged. This way if leader does not response we won't keep trying. Here is the comment about this:

```c++
    // We allow device to request router role upgrade using status
    // reason `kBorderRouterRequst` once while its local network data
    // remains unchanged. This ensures if the leader is running an
    // older version of Thread stack which does not support
    // `kBorderRouterRequst` reason, we do not keep trying (on no
    // response). The boolean `mDidRequestRouterRoleUpgrade` tracks
    // this. It is set to `false` when local network data get changed
    // or when partition ID gets changed (indicating a potential
    // leader change).
```

I think @EskoDijk described leader behavior nicely. 😃 
The OT code before https://github.com/openthread/openthread/pull/7578 would simply ignore the request (no response would be sent and reading the `Coap` code, I think no CoAP ack either).  But from sender side, the request will be timed out after some number of attempts (by `Coap` layer). Since we now limit this to one request, it should be ok to try once and time out.

simonlingoogle replied:
SGTM. Sending once after local network data change should work fine.

simonlingoogle reviewed (approved):
LGTM 👍

EskoDijk commented on src/core/thread/mle_router.cpp:
```diff
@@ -1438,7 +1438,11 @@ Error MleRouter::HandleAdvertisement(RxInfo &aRxInfo)
-            HasOneNeighborWithComparableConnectivity(route, routerId))
+            HasOneNeighborWithComparableConnectivity(route, routerId)
+#if OPENTHREAD_CONFIG_BORDER_ROUTER_ENABLE && OPENTHREAD_CONFIG_BORDER_ROUTER_REQUEST_ROUTER_ROLE
+            && !Get<NetworkData::Notifier>().IsEligibleForRouterRoleUpgradeAsBorderRouter()
```
The check for 'IsEligibleForRouterRoleUpgradeAsBorderRouter()' will return false if the number of routers is equal to MAX_ROUTER.  In that particular case, then the BR will downgrade even though it's entitled to keep its state as Router.
In practice this condition will likely never occur. And perhaps it's even a good feature: e.g. in case of 2 BRs providing connectivity that are Router, and MAX_ROUTERS is reached, one of the two will downgrade to make some room for others to become Router if needed. So I would just leave it as is but good to be aware of.

abtink replied:
Yes. I agree. Thanks @EskoDijk.

EskoDijk reviewed (approved):
Did a quick check, it looks good overall (made 2 minor comments)

simonlingoogle commented on src/core/thread/network_data.cpp:
```diff
@@ -814,5 +814,39 @@ uint8_t NetworkData::CountBorderRouters(RoleFilter aRoleFilter) const
+    ExternalRouteConfig route;
+    OnMeshPrefixConfig  prefix;
+
+    while (GetNextExternalRoute(iterator, route) == kErrorNone)
```
Just to mention that the NAT64 prefix is also counted as BR, which I think is correct.

abtink replied:
Yes.  Any external route entry (no condition on flags) counts.

abtink commented:
In the last pushed commit, just rebased on `main` and added logic to count domain prefix entry as one of the conidtions ofr "BR providing external IP connectivity" (to match the update in SPEC-1033).

abtink commented on src/core/thread/network_data.cpp:
```diff
@@ -814,5 +814,34 @@ uint8_t NetworkData::CountBorderRouters(RoleFilter aRoleFilter) const
     return rlocsLength;
 }
 
+bool NetworkData::ContainsBorderRouterWithRloc(uint16_t aRloc16) const
```
I think we check and include it.
```c++
    while (GetNextOnMeshPrefix(iterator, prefix) == kErrorNone)
    {
        if ((prefix.mRloc16 == aRloc16) && prefix.mOnMesh && (prefix.mDefaultRoute || prefix.mDp))
        {
            ExitNow(contains = true);
        }
    }
```
Do we need any other check for this?

jwhui replied:
Ah, I missed that. Thanks.

jwhui commented on src/core/config/border_router.h:
```diff
@@ -55,6 +55,26 @@
+ * default route flag) and is acting as a REED is eligible to request a router role upgrade by sending an "Address
+ * Solicit" request to leader with status reason `BorderRouterRequest`. This reason is used when the number of active
+ * routers in the Thread mesh is above the threshold, and only if the number of existing eligible BRs (determined from
+ * the Thread Network Data) that are acting as router is less than two. This mechanism  allows up to two eligible
```
```suggestion
 * the Thread Network Data) that are acting as router is less than two. This mechanism allows up to two eligible
```

jwhui commented on src/core/thread/mle_types.hpp:
```diff
@@ -162,6 +162,13 @@ constexpr uint8_t kRouterSelectionJitter      = 120; ///< (in sec)
 constexpr uint8_t kRouterUpgradeThreshold   = 16;
 
+/**
+ * Threshold to accept a router upgrade request with reason `kBorderRouterRequst` (number of BRs acting as router in
```
```suggestion
 * Threshold to accept a router upgrade request with reason `kBorderRouterRequest` (number of BRs acting as router in
```

jwhui commented on src/core/config/border_router.h:
```diff
@@ -55,6 +55,26 @@
+ * This config is applicable on an `OPENTHREAD_FTD` build and when `OPENTHREAD_CONFIG_BORDER_ROUTER_ENABLE` is also
+ * enabled.
+ *
+ * A Border Router which provides IP connectivity (either adding an external route prefix or an on-mesh prefix with
```
Mention domain prefix?

jwhui commented on src/core/thread/network_data.hpp:
```diff
@@ -325,9 +325,9 @@ class NetworkData : public InstanceLocator
+     * This method finds and returns the list of RLOCs of border routers providing external IP connectivity.
      *
-     * A border router is considered to provide external IPv6 connectivity if it has added at least one external route
+     * A border router is considered to provide external IP connectivity if it has added at least one external route
```
Mention domain prefix?

jwhui commented on src/core/thread/network_data_notifier.cpp:
```diff
@@ -158,6 +182,90 @@ void Notifier::HandleCoapResponse(Error aResult)
+
+    // Check that we are a border router providing IP connectivity and already
+    // in the leader's network data and therefore eligible to request router
+    // role upgrade with `kBorderRouterRequst` status.
```
```suggestion
    // role upgrade with `kBorderRouterRequest` status.
```

jwhui commented on src/core/thread/network_data_notifier.cpp:
```diff
@@ -158,6 +182,90 @@ void Notifier::HandleCoapResponse(Error aResult)
+void Notifier::ScheduleRouterRoleUpgradeIfEligible(void)
+{
+    // We allow device to request router role upgrade using status
+    // reason `kBorderRouterRequst` once while its local network data
```
```suggestion
    // reason `kBorderRouterRequest` once while its local network data
```

jwhui commented on src/core/thread/network_data_notifier.cpp:
```diff
@@ -158,6 +182,90 @@ void Notifier::HandleCoapResponse(Error aResult)
+    // reason `kBorderRouterRequst` once while its local network data
+    // remains unchanged. This ensures if the leader is running an
+    // older version of Thread stack which does not support
+    // `kBorderRouterRequst` reason, we do not keep trying (on no
```
```suggestion
    // `kBorderRouterRequest` reason, we do not keep trying (on no
```

jwhui commented on src/core/thread/network_data_notifier.hpp:
```diff
@@ -74,10 +76,29 @@ class Notifier : public InstanceLocator, private NonCopyable
+#if OPENTHREAD_FTD && OPENTHREAD_CONFIG_BORDER_ROUTER_ENABLE && OPENTHREAD_CONFIG_BORDER_ROUTER_REQUEST_ROUTER_ROLE
+    /**
+     * This method indicates whether the device as border router is eligible for router role upgrade request using the
+     * status reason `kBorderRouterRequst`.
```
```suggestion
     * status reason `kBorderRouterRequest`.
```

jwhui commented on src/core/thread/network_data_notifier.cpp:
```diff
@@ -158,6 +182,90 @@ void Notifier::HandleCoapResponse(Error aResult)
+        {
+            LogInfo("Requesting router role as BR");
+            mDidRequestRouterRoleUpgrade = true;
+            IgnoreError(Get<Mle::MleRouter>().BecomeRouter(ThreadStatusTlv::kBorderRouterRequst));
```
```suggestion
            IgnoreError(Get<Mle::MleRouter>().BecomeRouter(ThreadStatusTlv::kBorderRouterRequest));
```

jwhui commented on src/core/thread/network_data_notifier.hpp:
```diff
@@ -74,10 +76,29 @@ class Notifier : public InstanceLocator, private NonCopyable
+     * This method indicates whether the device as border router is eligible for router role upgrade request using the
+     * status reason `kBorderRouterRequst`.
+     *
+     * This method checks whether device is providing IP connectivity (adding an external router or an on-mesh prefix
```
Mention domain prefix?
